modules/afmongodb/afmongodb-legacy-uri.c

```diff
@@ -0,0 +1,289 @@
+                          NULL);
+              g_free(host);
+            }
+        }
+      else
+        {
+          gchar *localhost = g_strdup_printf(DEFAULTHOST ":%d", MONGOC_DEFAULT_PORT);
+          self->servers = g_list_append(NULL, localhost);
+          _append_host(&self->recovery_cache, DEFAULTHOST, MONGOC_DEFAULT_PORT);
+        }
+
+      if (!_parse_addr(g_list_nth_data(self->servers, 0), &self->address, &self->port))
+        {
+          msg_error("Cannot parse the primary host",
+                    evt_tag_str("primary", g_list_nth_data(self->servers, 0)),
+                    evt_tag_str("driver", self->super.super.super.id),
+                    NULL);
+          return FALSE;
+        }
+    }
+  else
+    {
+      if (!self->address)
+        {
+          msg_error("Cannot parse address",
+                    evt_tag_str("primary", g_list_nth_data(self->servers, 0)),
+                    evt_tag_str("driver", self->super.super.super.id),
+                    NULL);
+          return FALSE;
+        }
+      _append_host(&self->recovery_cache, self->address, 0);
+    }
+  return TRUE;
+}
+
+static gboolean
+_append_servers(GString *uri_str, const GList *recovery_cache, gboolean *have_uri)
+{
+  const GList *iterator = recovery_cache;
+  *have_uri = FALSE;
+  gboolean have_path = FALSE;
+  do
+    {
+      const MongoDBHostPort *hp = (const MongoDBHostPort *)iterator->data;
+      if (hp->port)
+        {
+          *have_uri = TRUE;
+          if (have_path)
+            {
+              msg_warning("Cannot specify both a domain socket and address", NULL);
+              return FALSE;
+            }
+          g_string_append_printf(uri_str, "%s:%d", hp->host, hp->port);
+        }
+      else
+        {
+          have_path = TRUE;
+          if (*have_uri)
+            {
+              msg_warning("Cannot specify both a domain socket and address", NULL);
+              return FALSE;
+            }
+          g_string_append_printf(uri_str, "%s", hp->host);
+        }
+      iterator = iterator->next;
+      if (iterator)
+        g_string_append_printf(uri_str, ",");
+    }
+  while (iterator);
+  return TRUE;
+}
+
+static gboolean
+_check_auth_options(MongoDBDestDriver *self)
+{
+  if (self->user || self->password)
+    {
+      if (!self->user || !self->password)
+        {
+          msg_error("Neither the username, nor the password can be empty", NULL);
+          return FALSE;
+        }
+    }
+  return TRUE;
+}
+
+gboolean
+afmongodb_dd_create_uri_from_legacy(MongoDBDestDriver *self)
+{
+  if (!_check_auth_options(self))
+    return FALSE;
+
+  if (self->uri_str && self->is_legacy)
+    {
+      msg_error("Error: either specify a MongoDB URI (and optional collection) or only legacy options",
+                evt_tag_str("driver", self->super.super.super.id),
+                NULL);
+      return FALSE;
+    }
+  else if (self->is_legacy)
```

Comment:
Maybe this block could be factored out to a private `_build_uri_from_legacy_options(MongoDBDestDriver *self)`.
The `else` branch is not necessary as we return from the `if`.

The final result could be something similar then:

```C
gboolean
afmongodb_dd_create_uri_from_legacy(MongoDBDestDriver *self)
{
  if (!_check_auth_options(self))
    return FALSE;

  if (_is_legacy_options_mixed_with_uri)
    return FALSE;

  return _build_uri_from_legacy_options(self);
}
```